utils/bounded_workers.go

```diff
@@ -0,0 +1,86 @@
+// (c) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package utils
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+type BoundedWorkers struct {
+	workerSpawner      chan struct{}
+	workerCount        atomic.Int32
+	outstandingWorkers sync.WaitGroup
+
+	outstandingWork sync.WaitGroup
+
+	work      chan func()
+	workClose sync.Once
+}
+
+// NewBoundedWorkers returns an instance of [BoundedWorkers] that
+// will spawn up to [max] goroutines.
+func NewBoundedWorkers(max int) *BoundedWorkers {
+	return &BoundedWorkers{
+		work:          make(chan func()),
+		workerSpawner: make(chan struct{}, max),
+	}
+}
+
+// startWorker creates a new goroutine to execute [f] immediately and then keeps the goroutine
+// alive to continue executing new work.
+func (b *BoundedWorkers) startWorker(f func()) {
+	b.workerCount.Add(1)
+	b.outstandingWorkers.Add(1)
+
+	go func() {
+		defer b.outstandingWorkers.Done()
+
+		if f != nil {
+			f()
+			b.outstandingWork.Done()
+		}
+		for f := range b.work {
+			f()
+			b.outstandingWork.Done()
+		}
+	}()
+}
+
+// Execute the given function on an existing goroutine waiting for more work, a new goroutine,
+// or return if the context is canceled.
+//
+// Execute must not be called after Wait, otherwise it might panic.
+func (b *BoundedWorkers) Execute(f func()) {
+	b.outstandingWork.Add(1)
+
+	// Ensure we feed idle workers first
+	select {
+	case b.work <- f:
+		return
+	default:
+	}
+
+	// Fallback to waiting for an idle worker or allocating
+	// a new worker (if we aren't yet at max concurrency)
+	select {
+	case b.work <- f:
+	case b.workerSpawner <- struct{}{}:
+		b.startWorker(f)
+	}
+}
+
+// Wait returns after all enqueued work finishes and all goroutines to exit.
+// Wait returns the number of workers that were spawned during the run.
+//
+// It is safe to call Wait multiple times but not safe to call [Execute]
+// after [Wait] has been called.
+func (b *BoundedWorkers) Wait() int {
+	b.outstandingWork.Wait()
```

Comment:
I don't think `outstandingWork` is necessary. If we wait for the workers to complete, then all of the work has been completed assuming:

Every call to `Execute` has completed before we call `Wait`, which we already assume to prevent a panic.

If this is the case, then all of the work is either completed already (executed synchronously on the calling thread) or has been added to the `work` channel with at least one worker created.

Since workers will not terminate until the work channel is closed, as soon as the workers have terminated, we should be guaranteed that all of the work has finished.

Reply:
You are right. This is a remnant of a more complex implementation I used to have where I allowed early termination.